06_gpu_and_ml/unsloth_finetune.py

```diff
@@ -0,0 +1,562 @@
+# This example demonstrates using Unsloth to finetune a version of Qwen3-14B with the
+# FineTome-100k dataset on Modal using only a single GPU!
+
+# To run this example, checkout the examples repo, install the Modal client, and run
+# `modal run -m 06_gpu_and_ml.unsloth_finetune`. You can also customize the training
+# process by tweaking hyperparameters with command line flags, e.g.
+# `modal run -m 06_gpu_and_ml.unsloth_finetune --max-steps 10000'.
+
+# ## Modal Infrastructure Setup
+
+import pathlib
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Optional
+
+import modal
+
+# We create a Modal [App](https://modal.com/docs/guide/apps) to organize our functions
+# and shared infrastructure like container images and volumes.
+
+app = modal.App("unsloth-finetune")
+
+# ### Container Image Configuration
+
+# We build a custom container image with Unsloth and all necessary dependencies.
+# The image includes the latest version of Unsloth (as of writing) with optimizations
+# for the latest model architectures. Once the image is defined, we can specify the
+# imports we'll need to write the rest of our training code. Importantly, we import
+# `unsloth` before the rest so that Unsloth's patches are applied to packages like
+# `transformers`, `peft`, and `trl`.
+
+train_image = (
+    modal.Image.debian_slim(python_version="3.11")
+    .uv_pip_install(
+        "accelerate==1.9.0",
+        "datasets==3.6.0",
+        "hf-transfer==0.1.9",
+        "huggingface_hub==0.34.2",
+        "peft==0.16.0",
+        "transformers==4.54.0",
+        "trl==0.19.1",
+        "unsloth[cu128-torch270]==2025.7.8",
+        "unsloth_zoo==2025.7.10",
+        "wandb==0.21.0",
+    )
+    .env({"HF_HOME": "/model_cache"})
+)
+
+with train_image.imports():
+    # unsloth must be first!
+    import unsloth  # noqa: F401,I001
+    import datasets
+    import torch
+    import wandb
+    from transformers import TrainingArguments
+    from trl import SFTTrainer
+    from unsloth import FastLanguageModel
+    from unsloth.chat_templates import standardize_sharegpt
+
+# ### Volume Configuration
+
+# Modal [Volumes](https://modal.com/docs/guide/volumes) provide persistent storage
+# that persists between function invocations. We use separate volumes for different
+# types of data to enable efficient caching and sharing.
+
+# Cache for downloaded model weights - shared across all experiments
+model_cache_volume = modal.Volume.from_name(
+    "unsloth-model-cache", create_if_missing=True
+)
+
+# Cache for processed datasets - shared when using the same dataset
+dataset_cache_volume = modal.Volume.from_name(
+    "unsloth-dataset-cache", create_if_missing=True
+)
+
+# Storage for training checkpoints and final models
+checkpoint_volume = modal.Volume.from_name(
+    "unsloth-checkpoints", create_if_missing=True
+)
+
+# ### Picking a GPU
+
+# We use L40S for its healthy balance of VRAM, CUDA cores, and clock speed.
```

Comment:
Nice-to-have: link out to the GPU Glossary for GPU terms like [VRAM](https://modal.com/gpu-glossary/device-hardware/gpu-ram) and [CUDA Core](https://modal.com/gpu-glossary/device-hardware/cuda-core)